Aftership

Jngai opened the pull request:
@cantino The spec is not ready but the agent is. I was a bit into this before I realized that you need a paying plan for the api. I can probably do more api integration with aftership but I think I should stop here. Its working but I would love some input.

Jngai commented:
I tested it locally as well.

Jngai commented:
@cantino The spec is good as well. I just put it in earlier.

cantino commented on app/models/agents/aftership_agent.rb:
```diff
@@ -0,0 +1,158 @@
+      You can use this agent to retrieve tracking data. You have to provide a specific `get` request and its associated option.
+ 
+      To get all trackings for your packages please enter `get` for key and `trackings` for the option.
+      To get tracking for a specific tracking number, add the extra keys `slug`, `tracking_number` and its associated values.
```
", and *their* associated values."

cantino commented on app/models/agents/aftership_agent.rb:
```diff
@@ -0,0 +1,158 @@
+      To get all trackings for your packages please enter `get` for key and `trackings` for the option.
+      To get tracking for a specific tracking number, add the extra keys `slug`, `tracking_number` and its associated values.
+
+      Set `single_tracking_request` to true as well.
```
What do you mean by this?  For the "specific tracking number" case?

cantino commented on app/models/agents/aftership_agent.rb:
```diff
@@ -0,0 +1,158 @@
+    end
+
+    def event_url
+      base_url + "#{URI.encode(interpolated[:get].to_s)}"
```
Maybe the `:get` option should be called `:path`?  Or `:endpoint`?

dsander commented on app/models/agents/aftership_agent.rb:
```diff
@@ -0,0 +1,158 @@
+    event_description <<-MD
+      A typical tracking event has 3 objects (attributes, tracking, and checkpoint) and it looks like this
+          {
+        "meta": {
```
The indentation looks a bit off here, I think the complete hash needs to be indented with 4 spaces so markdown renders it as a block like so?
<img width="541" alt="screenshot 2016-03-16 09 36 39" src="https://cloud.githubusercontent.com/assets/20943/13806269/dfde70d0-eb5a-11e5-90a3-e63596e9aa89.png">

Jngai commented:
@cantino ok I think I am good. I tested all 3 endpoints locally and addressed the feedbacks. I also removed a bug and put in something that I left out.

dsander commented on app/models/agents/aftership_agent.rb:
```diff
@@ -0,0 +1,156 @@
+
+    event_description <<-MD
+      A typical tracking event has 3 objects (attributes, tracking, and checkpoint) and it looks like this
+    {
```
Sorry, I meant 4 spaces further then the line before (you might also need a empty line in between). You can check the rendered markdown by creating a new agent and selecting the `AftershipAgent` as a source.

Jngai replied:
Its rendered correctly now but I removed the attributes object because its not returning anything useful. Compared to other agents, its a bit on the long side. Hope this works out.

cantino commented on app/models/agents/aftership_agent.rb:
```diff
@@ -0,0 +1,139 @@
+      To get the last checkpoint of a package set key to `path` and option to `last_checkpoint`. Please provide `slug` and `tracking_number`. Set `last_checkpoint_request` to true.
+
+      `slug` is a unique courier code. 
+
```
I'm a bit confused about the relationship of `single_tracking_request` and `last_checkpoint_request`, and when you use them.  Perhaps it would be cleaner to just have people provide the full path, with a link to the API?

All trackings:

```json
{
  'path': '/tracking',
}
```

Specific:

```json
{
  'path': '/tracking'/fedex/123,
}
```

checkpoint:

```json
{
  'path': '/last_checkpoint/fedex/123'
}
```

Then the docs could say:

Provide the path for the API endpoint that you'd like to hit.  For example, for all active packages, enter `/trackings` (see https://www.aftership.com/docs/api/4/trackings), for a specific package, use `/trackings/SLUG/TRACKING_NUMBER` and replace `SLUG` with a courier code and `TRACKING_NUMBER` with the tracking number, or use Liquid (with a link).

Whatcha think?

Jngai commented:
@cantino I understand. I wasn't sure how to explain the agent. Once I put your suggestions in I removed the unnecessary methods and modified the spec accordingly.

Jngai commented:
just double checked this locally

cantino commented:
Looks great, thanks @Jngai!
